### PR TITLE
Auto-format micromega files in pre-commit hook.

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** infrastructure / fix.

For a few months now, formatting of OCaml files in the micromega plugin has been checked in the linter thanks to `ocamlformat`, but the pre-commit hook did not automatically fix those issues. This commit fixes this by explicitly calling `ocamlformat` on relevant files (modified ml files). Note that when we get rid of the legacy build system, we will be able to use `dune build @fmt --auto-promote` instead of having to explicitly call `ocamlformat`. ~~Until then, we will have to manually keep in sync the list in `.ocamlformat-ignore` and what is checked in the pre-commit hook.~~

I don't know how to check that a file was updated, to print a message (or even the diff) if that was the case.

### CODE REVIEW

File: dev/tools/pre-commit
Comment (SkySkimmer) [inline dev/tools/pre-commit]:
```suggestion
if command -v ocamlformat > /dev/null
```

File: dev/tools/pre-commit
Comment (SkySkimmer) [inline dev/tools/pre-commit]: Unrestricted `git add -u` is no good
Possibly instead we should call ocamlformat next to `check-eof-newline.sh --fix` (and so call it twice)
We can blindly call it on all modified .ml(i) files, the ocamlformat-ignore will restrict effects appropriately.

Comment (Zimmi48) [reply dev/tools/pre-commit]: I tried to apply your suggestion but did not manage to reach a working state. Have a look at my attempted patch:

```diff
diff --git a/dev/tools/pre-commit b/dev/tools/pre-commit
index 028269b19c..feb2e9dd00 100755
--- a/dev/tools/pre-commit
+++ b/dev/tools/pre-commit
@@ -7,7 +7,17 @@ set -e
 
 dev/tools/check-overlays.sh
 
+# Can we check and fix formatting?
+if ! command -v ocamlformat > /dev/null
+then
+    1>&2 echo "Warning: ocamlformat is not in path. Cannot check formatting."
+    function ocamlformat() {
+        :
+    }
+fi
+
 if ! git diff --cached --name-only -z | xargs -0 dev/tools/check-eof-newline.sh ||
+        ! git diff --cached --name-only -z | grep -E '.*\.mli?$' | xargs -0 ocamlformat --check ||
         ! git diff-index --check --cached HEAD >/dev/null 2>&1 ;
 then
     1>&2 echo "Auto fixing whitespace issues..."
@@ -36,6 +46,7 @@ then
     git apply --cached --whitespace=fix "$index"
     git apply --whitespace=fix "$index" 2>/dev/null # no need to repeat yourself
     git diff --cached --name-only -z | xargs -0 dev/tools/check-eof-newline.sh --fix
+    git diff --cached --name-only -z | grep -E '.*\.mli?$' | xargs -0 ocamlformat -i
     git add -u
     1>&2 echo #newline
 
@@ -73,12 +84,3 @@ then
     # clean up temporary files
     rm "$index" "$tree" "$fixed_index"
 fi
-
-# Check and fix formatting
-if command -v ocamlformat > /dev/null
-then
-    ocamlformat -i $(find plugins/micromega/*.ml{,i} ! -name micromega.ml)
-    git add -u
-else
-    1>&2 echo "Warning: ocamlformat is not in path. Could not check formatting."
-fi
```

Comment (SkySkimmer) [reply dev/tools/pre-commit]: You probably need -z in your greps
Also you can't use bash functions with xargs, instead do
~~~bash
ocamlformat=$(command -v ocamlformat || echo true)
~~~
then call `xargs -0 ocamlformat ...`

Comment (Zimmi48) [reply dev/tools/pre-commit]: Thanks for your advice. This seems to have worked (see new commit). But there is something weird which is that formatting issues are restored after committing (even if they are fixed in the commit).

File: dev/tools/pre-commit
Comment (SkySkimmer) [inline dev/tools/pre-commit]: Add an ocamlformat call just above here

File: dev/tools/pre-commit
Comment (SkySkimmer) [inline dev/tools/pre-commit]: I'd do `[[ $ocamlformat = true ]]` or similar (feels easier to understand what's going on)

File: dev/tools/pre-commit
Comment (Zimmi48) [reply dev/tools/pre-commit]: Thanks, I should have realized. I was blinded by the fact that you initially told me that we would need two calls when in fact we needed three (if you count the `--check` in the test).

File: dev/tools/pre-commit
Comment (SkySkimmer) [inline dev/tools/pre-commit]:
```suggestion
if [ "$ocamlformat" = true ]
```

**Comment by SkySkimmer** [comment]
I guess the hook will fail if we try to commit partial work (with broken syntax)
Not sure how much this matters, cc @coq/pushers

**Comment by Zimmi48** [comment]
> I guess the hook will fail if we try to commit partial work (with broken syntax)

Only when doing so in files where formatting is enforced (only micromega for now and possibly funind soon, cf. #11776).

**Comment by ppedrot** [comment]
Beotian question: why is so special about micromega and funind that we want to automatically apply ocamlformat to them? They used to have horrible layout, but after one application of ocamlformat it should be fine now, shouldn't it?

**Comment by ejgallego** [comment]
> They used to have horrible layout, but after one application of ocamlformat it should be fine now, shouldn't it?

Indeed the idea was to apply it once; but it was pointed out that it would be good to prevent the style diverging again. Personally I'm split about this, but I see the point tho, specially in parts of the code that don't get many commits [for example it took 3 months for someone to have a problem with micromega]

After working in other projects that enable ocamlformat by default, at first I felt a bit frustrated on not having control over code formatting; but now it seems more like a liberation as I don't have to worry about style anymore.

**Comment by Zimmi48** [comment]
The main point for micromega was to allow its main maintainer to not have to worry about formatting and avoid getting review comments about this.

>after one application of ocamlformat it should be fine now, shouldn't it?

The same could have been said of trailing whitespace. And yet, they were largely re-introduced after the big commit that removed them. Bad formatting can be re-introduced too.

> After working in other projects that enable ocamlformat by default, at first I felt a bit frustrated on not having control over code formatting; but now it seems more like a liberation as I don't have to worry about style anymore.

This is the usual story of people learning to work with a formatter. OCaml is an old language, so it will take time until its programmers learn that this is best to work with a formatter rather than not, but programmers of modern languages like Go, Elm and ReasonML know this already.

So now that the pre-commit hook is fixed to apply ocamlformat, the debate (if there must be a debate) should not be on its usefulness or not but on which style we want to have (for now, the settings were not really discussed).

**Comment by gares** [comment]
> OCaml is an old language, so it will take time until its programmers learn that this is best to work with a formatter rather than not

For "old" programmers it would help to have a list of benefits rather than a statement about their age.

For example I think it takes zero time to convince a C programmer that automatic memory management is useful, because it can clearly see the benefits. Indeed, in their experience, memory bugs are many and nasty.

If does not take zero time to convince "old" programmer that they are in desperate need code formatting, it may be because that they may fail too see the benefits.

**Comment by maximedenes** [comment]
> I guess the hook will fail if we try to commit partial work (with broken syntax)
> Not sure how much this matters, cc @coq/pushers

This is close to a deal breaker to me. I often do dirty WIP commits (basically a few times a day, when I interrupt my work). Please fix it before merging this PR.

**Comment by Zimmi48** [comment]
> If does not take zero time to convince "old" programmer that they are in desperate need code formatting, it may be because that they may fail too see the benefits.

My observation is that it takes a lot of effort to convince experienced programmers that they should use a code formatter, but the benefit is something that you learn by yourself while using it. I cannot do more than just quote Emilio's testimony.

OTOH in this case, we are only talking about micromega (and maybe funind too). This is a pretty reasonably delimited sphere of application. And it is only introduced if the maintainers of this component approve it. So I don't think that it is worth debating whether code formatting is beneficial with the whole team (*as long as it does not make anyone's life harder*).

> This is close to a deal breaker to me. I often do dirty WIP commits (basically a few times a day, when I interrupt my work). Please fix it before merging this PR.

OK

**Comment by cpitclaudel** [comment]
> This is close to a deal breaker to me. I often do dirty WIP commits (basically a few times a day, when I interrupt my work).

I think you can skip pre-commit hooks with `git commit --no-verify`?

**Comment by SkySkimmer** [comment]
>This is close to a deal breaker to me. I often do dirty WIP commits (basically a few times a day, when I interrupt my work). Please fix it before merging this PR.

I guess we can just ignore errors from ocamlformat (`|| true` after the calls)

**Comment by Zimmi48** [comment]
> > This is close to a deal breaker to me. I often do dirty WIP commits (basically a few times a day, when I interrupt my work). Please fix it before merging this PR.
> 
> I guess we can just ignore errors from ocamlformat (`|| true` after the calls)

But it would still print "Auto fixing whitespace and formatting issues..."

**Comment by Zimmi48** [comment]
Ideally, I would have wanted to use an option of `ocamlformat` which checks that the syntax is correct but it does not exist and the option `--check` gives the same exit code if the formatting is incorrect and if the syntax is incorrect. The only different is that it is silent in the first case and verbose in the second.

**Comment by JasonGross** [comment]
> > > This is close to a deal breaker to me. I often do dirty WIP commits (basically a few times a day, when I interrupt my work). Please fix it before merging this PR.
> > 
> > 
> > I guess we can just ignore errors from ocamlformat (`|| true` after the calls)
> 
> But it would still print "Auto fixing whitespace and formatting issues..."

Perhaps instead do `|| echo "ocamlformat failed with $?"` or something?

**Comment by SkySkimmer** [comment]
>But it would still print "Auto fixing whitespace and formatting issues..."

That's fine, it fixes what it can and if the commit is amended to fix syntax errors it should fix the rest.

**Comment by Zimmi48** [comment]
I found another (and IMO better) way. Please check the last commit.

File: dev/tools/pre-commit
Comment (SkySkimmer) [inline dev/tools/pre-commit]: This seems dangerous, I suspect that if we do
- commit a.ml and b.ml with b.ml having syntax error
- amend with fix to b.ml (not touching a.ml)

then the formatting for a.ml will not be fixed.

Why do you think ignoring errors from ocamlformat is worse?

Comment (Zimmi48) [reply dev/tools/pre-commit]: Indeed, but that's the problem of the people whose commits do not compile. This is something that will be rejected by the linter anyway. Am I correct?

>Why do you think ignoring errors from ocamlformat is worse?

To get clear error messages. With your solution, you could get "Fixing..." even if nothing is fixed, whitespace nor formatting.

**Comment by ejgallego** [comment]
> Ideally, I would have wanted to use an option of `ocamlformat` which checks that the syntax is correct but it does not exist

Maybe open a feature request upstream?

File: dev/tools/pre-commit
Comment (SkySkimmer) [reply dev/tools/pre-commit]: >To get clear error messages. With your solution, you could get "Fixing..." even if nothing is fixed, whitespace nor formatting.

Change it to "attempting to fix", problem solved?

Comment (Zimmi48) [reply dev/tools/pre-commit]: I don't understand what is the issue with my solution and the fact that it is anyway a very edge case and not a big deal that some formatting is not fixed in this case.

Comment (SkySkimmer) [reply dev/tools/pre-commit]: >I don't understand what is the issue with my solution

We know how to do better.

Another scenario:
- `git add` a file with valid syntax
- break the file in the worktree
- `git commit`

The commit will have bad formatting.

Comment (Zimmi48) [reply dev/tools/pre-commit]: OK, thanks, I understand the problem better with this scenario.

Comment (Zimmi48) [reply dev/tools/pre-commit]: Actually @SkySkimmer the problem you describe in your scenario exists even without this preliminary check. Suppose I `git add` a file with broken syntax, then fix it in the worktree (including formatting), then `git commit` without option `-a`. The pre-commit hook will skip the fixing step and thus my commit will have bad formatting.

File: dev/tools/pre-commit
Comment (SkySkimmer) [inline dev/tools/pre-commit]: >

Actually @SkySkimmer the problem you describe in your scenario exists even without this preliminary check. Suppose I git add a file with broken syntax, then fix it in the worktree (including formatting), then git commit without option -a. The pre-commit hook will skip the fixing step and thus my commit will have bad formatting.

True, I guess we should just remove these tests and run unconditionally.

Comment (Zimmi48) [reply dev/tools/pre-commit]: This PR is taking me far beyond what I was expecting. I was just hoping to help by providing a quick patch over the weekend. Would you like to finish it?

Comment (SkySkimmer) [reply dev/tools/pre-commit]: OK

**Comment by SkySkimmer** [comment]
I guess this needs a new assignee @coq/dev-tools-maintainers

**Comment by Zimmi48** [comment]
Is this OK that now every commit will print

```
Auto fixing whitespace and formatting issues...
Patches are saved in 'git-fix-ws-index.7xhlSC', 'git-fix-ws-index-fixed.r2KNDz' and 'git-fix-ws-tree.PGhLE1'.
If an error destroys your changes you can recover using them.
(The files are cleaned up on success.)

Fixing staged changes...

Fixing unstaged changes...

Whitespace and formatting pass complete.
```

?

**Comment by SkySkimmer** [comment]
>Is this OK that now every commit will print

It should be fine IMO

**Review by ejgallego — APPROVED** [review]
LGTM will leave one day for comments; it seems to me that this PR goes a bit beyond what the PR title states tho; I dunno if you would like to update it.

**Pull request merged**
